### PR TITLE
[PPP-3506][PPP-3533] pentaho-kettle: XXE Security Vulnerability in xml parsers

### PR DESCRIPTION
**Pull request by IvanNikolaychuk**
- Applying security to parsing, centralising xml-factory creation in XMLUtils class.
- Tests written
- Checkstyle applied

### CODE REVIEW

**Comment by IvanNikolaychuk** [comment]
@mchen-len-son , could you please review it please?

File: core/test-src/org/pentaho/di/core/xml/XMLHandlerUnitTest.java [hunk 1]
```diff
@@ -22,74 +22,79 @@
...
   public void openTagWithExternalBuilder() {
     StringBuilder builder = new StringBuilder( "qwe" );
-    openTag( builder, "rty" );
+    XMLHandler.openTag( builder, "rty" );
     assertEquals( "qwe<rty>", builder.toString() );
   }
 
   @Test
   public void closeTagWithNotNull() {
-    assertEquals( "</qwerty>", closeTag( "qwerty" ) );
```
Comment (matthewtckr) [inline core/test-src/org/pentaho/di/core/xml/XMLHandlerUnitTest.java]: Were these tests removed because they were no longer valid?

Comment (IvanNikolaychuk) [inline core/test-src/org/pentaho/di/core/xml/XMLHandlerUnitTest.java]: Nope, they shouldn't be removed.
Didn't notice that, thanks for remark

**Comment by IvanNikolaychuk** [comment]
@mchen-len-son , could you please take a look at this PR?

**Comment by IvanNikolaychuk** [comment]
@lucboudreau , could you please restart wingman?
I'm a bit confused, as I don't see any reasons for him to fail.

**Comment by lucboudreau** [comment]
Looks like a classpath issue. 

```
[javac] /home/buildguy/project/engine/test-src/org/pentaho/di/repository/RepositoryExportSaxParserTest.java:40: error: package test.util does not exist
[javac] import test.util.XXEUtils;
[javac]                 ^
```

**Comment by IvanNikolaychuk** [comment]
@lucboudreau , thanks for your assistance!
Not sure either, will try to figure it out tommorow.

File: core/test-src/org/pentaho/di/core/xml/XMLHandlerUnitTest.java [hunk 2]
```diff
@@ -22,74 +22,97 @@
...
+      + " <!ENTITY lol3 \"&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;\">\n"
+      + " <!ENTITY lol4 \"&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;\">\n"
+      + " <!ENTITY lol5 \"&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;\">\n"
+      + " <!ENTITY lol6 \"&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;\">\n"
+      + " <!ENTITY lol7 \"&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;\">\n"
+      + " <!ENTITY lol8 \"&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;\">\n"
+      + " <!ENTITY lol9 \"&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;\">\n"
+      + "]>\n"
+      + "<lolz>&lol9;</lolz>";
+
```
Comment (mchen-len-son) [inline core/test-src/org/pentaho/di/core/xml/XMLHandlerUnitTest.java]: As I can see, you have malicious xml in XXEUtils, and this one can be replaced.

**Comment by IvanNikolaychuk** [comment]
MALICIOUS_XML is duplicated in several modules.
It is done consciously, as it is impossible to concentrate it in one place, without adding new test-dependencies among modules. 
I don't think that we need to do this in this case.

**Comment by mchen-len-son** [comment]
Double checked locally.
Checkstyle violations contain only legal static imports for tests.

Note: This might break the plugin build for a while, due to unpredictabe order of modules building.

Merging.

**Comment by lucboudreau** [comment]
@mchen-len-son @IvanNikolaychuk If this might break the builds, please get in touch with the build folks to make sure they are aware. Ideally, please talk to them before you merge these types of changes. cc. @buildguy

**Comment by lgrill-pentaho** [comment]
Since this is kettle, there will be many opportunities to break downstream builds that depend on it, but since there are no new/changed dependencies or build configurations here, this is the kind of thing we will need to shake out on devci and possibly QAT unfortunately.

I expect that our developers are more in tune with the downstream impacts of xml processing changes in kettle.

**Pull request merged**